internal/controller/numaflowcontroller/numaflowcontroller_controller.go

```diff
@@ -741,3 +789,84 @@ func (r *NumaflowControllerReconciler) areDependentResourcesDeleted(ctx context.
 
 	return false
 }
+
+// get the tag of the numaflow container
+func getControllerDeploymentVersion(deployment *appsv1.Deployment) (string, error) {
+
+	c, err := config.GetConfigManagerInstance().GetConfig()
+	if err != nil {
+		return "", fmt.Errorf("error getting ConfigMap: %w", err)
+	}
+	imageNames := []string{DefaultNumaflowControllerImageName}
+	if len(c.NumaflowControllerImageNames) > 0 {
+		imageNames = c.NumaflowControllerImageNames
+	}
+
+	// in case the Deployment has sidecars, find the container whose image is named "numaflow"
+	containers := deployment.Spec.Template.Spec.Containers
+	for _, c := range containers {
+		imageName := c.Image
+		tag := ""
+		colon := strings.Index(c.Image, ":")
+		if colon != -1 {
+			imageName = c.Image[0:colon]
+			tag = c.Image[colon+1:]
+		}
+		finalSlash := strings.LastIndex(imageName, "/")
+		if finalSlash != -1 {
+			imageName = imageName[finalSlash+1:]
+
+		}
+		// is this is the Numaflow Controller itself?
+		isNumaflowController := false
+		for _, nfControllerImageName := range imageNames {
+			if imageName == nfControllerImageName {
+				isNumaflowController = true
+				break
+			}
+		}
+		if isNumaflowController {
+			if tag == "" {
+				return "", fmt.Errorf("no tag found in image path %q from Deployment %+v", c.Image, deployment)
+			} else {
+				// expect first letter to be a "v"
+				if tag[0] != 'v' {
+					return "", fmt.Errorf("expected numaflow-controller tag to start with 'v': %q", tag)
+				}
+				return tag[1:], nil
+			}
+		}
+	}
+	return "", fmt.Errorf("couldn't find image named %q in Deployment %+v", DefaultNumaflowControllerImageName, deployment)
+}
+
+// deleteNumaflowControllerChildren deletes child resources associated with a NumaflowController
+// for a specified version and namespace. It determines the target objects to delete and attempts
+// to remove them using the Kubernetes client. If a resource is not found, it continues with the
+// next one. Returns an error if unable to determine target objects or delete a resource.
+func (r *NumaflowControllerReconciler) deleteNumaflowControllerChildren(controller *apiv1.NumaflowController, currentVersion, namespace string) error {
+
+	targetObjs, err := determineTargetObjects(controller, currentVersion)
+	if err != nil {
+		return fmt.Errorf("unable to determine the target objects for the current version %s: %w", currentVersion, err)
```

Comment:
How about if this were to happen, we handle it gracefully? Let's say that the current version's definition were for some reason no longer there

Reply:
Ideally, as a back up, determine the target objects from the new version?

Reply:
I assumed this would not be a scenario that could happen. I'll check the new version as backup though.